parsel/selector.py

```diff
@@ -267,34 +282,99 @@ def __init__(
         text: Optional[str] = None,
         type: Optional[str] = None,
         namespaces: Optional[Mapping[str, str]] = None,
-        root: Optional[Any] = None,
+        root: Optional[Any] = _NOTSET,
         base_url: Optional[str] = None,
         _expr: Optional[str] = None,
     ) -> None:
-        self.type = st = _st(type or self._default_type)
-        self._parser = _ctgroup[st]["_parser"]
-        self._csstranslator = _ctgroup[st]["_csstranslator"]
-        self._tostring_method = _ctgroup[st]["_tostring_method"]
+        if type not in ("html", "json", "text", "xml", None):
+            raise ValueError(f"Invalid type: {type}")
 
-        if text is not None:
-            if not isinstance(text, str):
-                msg = f"text argument should be of type str, got {text.__class__}"
-                raise TypeError(msg)
-            root = self._get_root(text, base_url)
-        elif root is None:
+        self._text = text
```

Comment:
What do you think about moving this assignment below the next two 'if' blocks, so that self._text is not assigned until we validated that parameters are correct?

Reply:
It could go even lower, next to `self._expr = _expr` to couple both set procedures, as the third 'if' block also doesn't use such property from the private attribute.